yeat/Snakefile

```diff
@@ -102,7 +102,7 @@ rule spades:
     threads: 128
     shell:
         """
-        spades.py -1 {input.read1} -2 {input.read2} -t {threads} -o analysis/spades
+        spades.py -1 {input.read1} -2 {input.read2} -t {threads} -o analysis/spades {config[extraargs][spades]}
```

Comment:
In the snakemake file, we then call the dictionary and retrieve the key's value based on assembly.

Because we do a hard check checking for valid assembly algorithms in the config file, I hardcoded the keys here for each assembly algorithm.

Reply:
You're on the right path here, and the hard-coding is appropriate. Honestly, this is a perfectly viable solution.

The only suggestion I would offer involves Snakemake `params`. Sometimes when bracketed values in my `shell` statements get too hairy, I pull that out into the rule's `params` block. That gives more flexibility (especially if function calls are needed) and keeps the `shell` statement clean. For this rule, it might look something like this.

```python
rule spades:
    output:
        contigs="analysis/spades/{sample}_contigs.fasta"
    input:
        read1="seq/downsample/{sample}.R1.fq.gz",
        read2="seq/downsample/{sample}.R2.fq.gz"
    threads: 128
    params:
        extra_args=config["extra_args"]["spades"]
    shell:
        """
        spades.py -1 {input.read1} -2 {input.read2} -t {threads} -o analysis/spades {params.extra_args}
        ln -s scaffolds.fasta {output.contigs}
        """
```